[JSC] Use C++20 Concepts more in JavaScriptCore

zakariaridouh opened the pull request:
#### 2055454bfac883a8a5288ad96fece0561b5b7732
<pre>
[JSC] Use C++20 Concepts more in JavaScriptCore
<a href="https://bugs.webkit.org/show_bug.cgi?id=302744">https://bugs.webkit.org/show_bug.cgi?id=302744</a>
<a href="https://rdar.apple.com/165000894">rdar://165000894</a>

Reviewed by Darin Adler.

Use C++ 20 concepts more in JSC.

* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
* Source/JavaScriptCore/runtime/WeakMapImpl.h:
(JSC::WeakMapBucket::setValue):
(JSC::WeakMapBucket::value const):
(JSC::WeakMapBucket::copyFrom):
(JSC::WeakMapBucket::offsetOfValue):
(JSC::WeakMapBucket::extractValue):
(JSC::WeakMapBucket::visitAggregate):
(JSC::WeakMapBucket::clearValue):
(JSC::WeakMapImpl::get):
(JSC::WeakMapImpl::getBucket):
(JSC::WeakMapImpl::isWeakMap):
(JSC::WeakMapImpl::isWeakSet):

Canonical link: <a href="https://commits.webkit.org/304276@main">https://commits.webkit.org/304276@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0a42cce0dbeb62a67f6f5dca52c0c6280196a684

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/135069 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7469 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46322 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142583 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86892 "Built successfully") | ⏳ 🛠 ios-apple 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/8099 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7318 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103217 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/70455 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138015 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5754 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/121060 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84070 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5569 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3179 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/3172 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/127089 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/114780 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/39219 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/145284 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/133565 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/7151 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39794 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/111594 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7198 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5998 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111958 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28410 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5397 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117346 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/61088 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7201 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35509 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/166444 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6967 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70770 "Built successfully") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/43499 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7197 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7073 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

colelao commented on Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
```diff
@@ -2757,8 +2761,8 @@ class LowerDFGToB3 {
-    template <typename Generator, typename Func1, typename Func2,
-        typename = std::enable_if_t<std::is_function<typename std::remove_pointer<Func1>::type>::value && std::is_function<typename std::remove_pointer<Func2>::type>::value>>
+    template<typename Generator, typename Func1, typename Func2>
+        requires (std::is_function_v<std::remove_pointer_t<Func1>> && std::is_function_v<std::remove_pointer_t<Func2>>)
```
You could turn this into a concept
```
template <typename T> concept Function = std::is_function_v<std::remove_pointer_t<Func1>>;
```
then use it like this
```
template <typename Generator, Function Func1, Function Func2>
void compileUnaryMathIC(UnaryArithProfile* arithProfile, Func1 repatchingFunction, Func2 nonRepatchingFunction) {
...
}
```

You could also reuse that concept below

colelao commented on Source/JavaScriptCore/runtime/WeakMapImpl.h:
```diff
@@ -67,32 +67,38 @@ class WeakMapBucket {
 
-    template <typename T = Data>
-    ALWAYS_INLINE typename std::enable_if<std::is_same<T, WeakMapBucketDataKeyValue>::value>::type setValue(VM& vm, JSCell* owner, JSValue value)
+    template<typename T = Data>
```
Ditto as before. 
 
These could be two different concepts i.e. `SameAsDataKeyValue` and `SameAsDataKey`. Then, these could be reused through this file

colelao reviewed (approved):
Just a few style things, but it looks good.

colelao reviewed (approved):
Great changes!

darinadler commented on Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
```diff
@@ -161,6 +161,9 @@ JSC_DEFINE_NOEXCEPT_JIT_OPERATION_WITH_ATTRIBUTES(ftlUnreachable, NO_RETURN_DUE_
 concept isB3SupportedExceptionOperationResult = isExceptionOperationResult<OperationResultType> && !std::is_floating_point_v<typename OperationResultType::ResultType>;
 
+template<typename T>
+concept Function = std::is_function_v<std::remove_pointer_t<T>>;
```
It’s a bit peculiar to have one concept named `isB3SupportedExceptionOperationResult` and another named `Function` in the same source file. Not really a consistent style in this file.

darinadler commented on Source/JavaScriptCore/runtime/WeakMapImpl.h:
```diff
@@ -56,6 +56,12 @@ struct WeakMapBucketDataKeyValue {
+concept SameAsDataKeyValue = std::same_as<T, WeakMapBucketDataKeyValue>;
+
+template<typename T>
+concept SameAsDataKey = std::same_as<T, WeakMapBucketDataKey>;
```
Given that these concepts are in the top level of the `JSC` namespace, we probably want to use the prefix `WeakMap` just the way the the types have `WeakMap` prefixes.

zakariaridouh committed:
[JSC] Use C++20 Concepts more in JavaScriptCore
https://bugs.webkit.org/show_bug.cgi?id=302744
rdar://165000894

Reviewed by Darin Adler.

Use C++ 20 concepts more in JSC.

* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
* Source/JavaScriptCore/runtime/WeakMapImpl.h:
(JSC::WeakMapBucket::setValue):
(JSC::WeakMapBucket::value const):
(JSC::WeakMapBucket::copyFrom):
(JSC::WeakMapBucket::offsetOfValue):
(JSC::WeakMapBucket::extractValue):
(JSC::WeakMapBucket::visitAggregate):
(JSC::WeakMapBucket::clearValue):
(JSC::WeakMapImpl::get):
(JSC::WeakMapImpl::getBucket):
(JSC::WeakMapImpl::isWeakMap):
(JSC::WeakMapImpl::isWeakSet):

Canonical link: https://commits.webkit.org/304276@main
